src/mount/lite.clj

```diff
@@ -190,25 +209,26 @@
        ~@body)))
 
 (defmacro with-session
-  "Creates a new thread, with a new system of states. All states are
-  initially in the stopped status in this thread, regardless of the
-  status in the thread that spawns this new session. This spawned
-  thread and its subthreads will automatically use the states that are
-  started within this thread or subthreads. Exiting the spawned thread
-  will automatically stop all states in this session.
-
-  Returns a map with the spawned :thread and a :promise that will be
-  set to the result of the body or an exception."
+  "Creates a new thread, with a new system of states. All states are initially
+  in the stopped status in this thread, regardless of the status in the thread
+  that spawns this new session. This spawned thread and its subthreads, futures,
+  and agents will automatically use the states that are started within this
+  thread or subthreads. Exiting the spawned thread will automatically stop all
+  states in this session.
+
+  Returns a map with the spawned :thread and a :promise that will be set to the
+  result of the body or an exception."
   [& body]
   `(let [p# (promise)]
      {:thread (doto (Thread. (fn []
-                               (.set ^InheritableThreadLocal @#'itl (Thread/currentThread))
-                               (try
-                                 (deliver p# (do ~@body))
-                                 (catch Throwable t#
-                                   (deliver p# t#)
-                                   (throw t#))
-                                 (finally
-                                   (stop)))))
+                               (binding [*session* (new-session)]
+                                 (.set ^InheritableThreadLocal itl *session*)
+                                 (try
+                                   (deliver p# (do ~@body))
+                                   (catch Throwable t#
+                                     (deliver p# t#)
+                                     (throw t#))
+                                   (finally
+                                     (stop))))))
```

Comment:
I think that this should be converted to a future, which can be derefed directly and is run within a thread implicitly. However, I didn't want to make a potentially breaking change without checking with you first, @aroemers.